File: server/typescript/packages/codegen-ts/test/golden/fr019-shared-provided-conformance.test.ts
// Cross-port FR-019 conformance (TS — the reference). Loads the shared fixture
// `fixtures/codegen-conformance/shared-provided-enum/input/meta.json` and asserts the two
// FR-019 behaviors (ADR-0026):
//   • Shared materialization — a package-level abstract `field.enum` (`Priority`) extended by
//     TWO entities is materialized ONCE in `enums.ts`; both entities reference it.
//   • `@provided` — a package-level abstract `field.enum` (`Currency`, `@provided: true`) is
//     NOT materialized; consumers import it from the configured `providedEnumModule`, while
//     `@values` still drive the z.enum validation + the DB CHECK.
//   • A `@provided` enum with no `providedEnumModule` config → a codegen-time error.

import { describe, test, expect } from "bun:test";
import { mkdtempSync, rmSync, readFileSync, readdirSync, existsSync } from "node:fs";
import { tmpdir } from "node:os";
import { join, dirname } from "node:path";
import { runGen, defineConfig } from "../../src/index.js";
import { entityFile, queriesFile, barrel } from "../../src/generators/index.js";
import { MetaDataLoader, InMemoryStringSource, type MetaRoot } from "@metaobjectsdev/metadata";

function findRepoRoot(start: string): string {
  let dir = start;
  while (true) {
    if (existsSync(join(dir, "fixtures")) && existsSync(join(dir, "server"))) return dir;
    const parent = dirname(dir);
    if (parent === dir) throw new Error("could not locate repo root from " + start);
    dir = parent;
  }
}

async function loadSharedFixture(): Promise<MetaRoot> {
  const fxDir = join(findRepoRoot(import.meta.dir), "fixtures", "codegen-conformance", "shared-provided-enum", "input");
  const sources = readdirSync(fxDir)
    .filter((f) => f.endsWith(".json"))
    .map((f) => new InMemoryStringSource(readFileSync(join(fxDir, f), "utf-8"), { id: f, format: "json" }));
  const res = await new MetaDataLoader().load(sources);
  expect(res.errors, "shared FR-019 fixture must load cleanly").toEqual([]);
  return res.root;
}

async function gen(root: MetaRoot, providedEnumModule?: string): Promise<Record<string, string>> {
  const tmp = mkdtempSync(join(tmpdir(), "fr019-conf-"));
  try {
    await runGen({
      config: defineConfig({
        outDir: tmp,
        extStyle: "none",
        dbImport: "~/db",
        dialect: "postgres",
        generators: [entityFile(), queriesFile(), barrel()],
        ...(providedEnumModule !== undefined && { providedEnumModule }),
      }),
      metadata: root,
    });
    const files: Record<string, string> = {};
    for (const name of readdirSync(tmp)) if (name.endsWith(".ts")) files[name] = readFileSync(join(tmp, name), "utf-8");
    return files;
  } finally {
    rmSync(tmp, { recursive: true, force: true });
  }
}

describe("FR-019 shared + provided enum conformance (TS)", () => {
  test("shared abstract enum is materialized ONCE; both consumers reference it", async () => {
    const files = await gen(await loadSharedFixture(), "@acme/ext-enums");
    const enums = files["enums.ts"]!;
    expect(enums, "shared enums.ts must be emitted").toBeDefined();
    // Materialized exactly once.
    expect(enums.split('export type Priority =').length - 1).toBe(1);
    expect(enums).toContain('export type Priority = "LOW" | "MEDIUM" | "HIGH";');
    // Both entities reference it, neither redeclares the union.
    for (const name of ["Ticket.ts", "Order.ts"]) {
      const f = files[name]!;
      expect(f).toBeDefined();
      expect(f).toContain('from "./enums"');
      expect(f).toContain("PriorityEnum");
      expect(f).not.toContain("export type Priority =");
    }
  });

  test("@provided enum is NOT materialized; referenced from the configured module; @values still drive validation + CHECK", async () => {
    const files = await gen(await loadSharedFixture(), "@acme/ext-enums");
    const enums = files["enums.ts"]!;
    // Currency is provided → never materialized.
    expect(enums).not.toContain("Currency");
    const t = files["Ticket.ts"]!;
    // Referenced from the configured external module.
    expect(t).toContain('export { type Currency } from "@acme/ext-enums";');
    // @values stay the SSOT: validation + CHECK are still emitted from the members.
    expect(t).toContain('z.enum(["USD", "EUR", "GBP"])');
    expect(t).toContain("currency IN ('USD', 'EUR', 'GBP')");
  });

  test("a @provided enum with no providedEnumModule config is a codegen-time error naming the enum", async () => {
    let err: unknown;
    try {
      await gen(await loadSharedFixture()); // no providedEnumModule
    } catch (e) {
      err = e;
    }
    expect(err).toBeInstanceOf(Error);
    expect((err as Error).message).toContain("Currency");
  });
});

File: server/java/codegen-spring/src/test/java/com/metaobjects/generator/spring/Fr019SharedProvidedConformanceTest.java
package com.metaobjects.generator.spring;

import com.metaobjects.loader.MetaDataLoader;
import com.metaobjects.registry.SharedRegistryTestBase;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.TemporaryFolder;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.HashMap;
import java.util.Map;

import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;
import static org.junit.Assert.fail;

/**
 * Cross-port FR-019 conformance (Java/Spring). Loads the shared fixture
 * {@code fixtures/codegen-conformance/shared-provided-enum/input/meta.json} and asserts the two
 * FR-019 behaviors (ADR-0026):
 * <ul>
 *   <li><b>Shared materialization</b> — a package-level abstract {@code field.enum}
 *       ({@code Priority}) extended by TWO entities is materialized ONCE as a standalone
 *       {@code Priority.java}; both DTOs reference it (no nested redeclaration).</li>
 *   <li><b>{@code @provided}</b> — a package-level abstract {@code field.enum} ({@code Currency},
 *       {@code @provided: true}) is NOT materialized; consuming DTOs reference it at the configured
 *       namespace ({@code <ns>.Currency}). A referenced provided enum with no configured namespace
 *       is a codegen-time error naming the enum.</li>
 * </ul>
 *
 * <p>Each port runs the SAME fixture (the oracle) and asserts its own idiomatic contract — no
 * byte-identical cross-language expectation.</p>
 */
public class Fr019SharedProvidedConformanceTest extends SharedRegistryTestBase {

    @Rule
    public TemporaryFolder tmp = new TemporaryFolder();

    private MetaDataLoader loadShared() throws IOException {
        Path repoRoot = Path.of(System.getProperty("user.dir")).resolve("../../..").normalize();
        Path fixture = repoRoot.resolve("fixtures/codegen-conformance/shared-provided-enum/input/meta.json");
        assertTrue("shared FR-019 fixture must exist at " + fixture, Files.exists(fixture));
        return SpringTestFixtures.loadFixture(tmp.newFolder().toPath(), "sharedProvidedEnum",
            Files.readString(fixture));
    }

    /** Run the DTO generator with the given optional {@code providedEnumNamespace} fallback. */
    private Path run(MetaDataLoader loader, String providedEnumNamespace) throws IOException {
        Path out = tmp.newFolder().toPath();
        Map<String, String> args = new HashMap<>();
        args.put("outputDir", out.toString());
        if (providedEnumNamespace != null) args.put("providedEnumNamespace", providedEnumNamespace);
        SpringDtoGenerator gen = new SpringDtoGenerator();
        gen.setArgs(args);
        gen.execute(loader);
        return out;
    }

    @Test
    public void sharedEnumIsMaterializedOnceAndReferencedByBothEntities() throws Exception {
        Path out = run(loadShared(), "com.acme.ext");

        // Materialized ONCE as a standalone top-level enum in the declaring package.
        Path sharedEnum = out.resolve("acme/shop/Priority.java");
        assertTrue("standalone Priority.java must be materialized; expected at " + sharedEnum,
            Files.exists(sharedEnum));
        assertTrue("standalone enum must declare members verbatim",
            Files.readString(sharedEnum).contains("public enum Priority { LOW, MEDIUM, HIGH }"));

        // Both entities reference it; neither redeclares the enum inline.
        for (String name : new String[] { "TicketDto.java", "OrderDto.java" }) {
            Path dto = out.resolve("acme/shop").resolve(name);
            assertTrue(name + " must be emitted", Files.exists(dto));
            String src = Files.readString(dto);
            assertTrue("priority component must be typed as the shared enum in " + name + "; saw:\n" + src,
                src.contains("Priority priority"));
            assertFalse("shared enum must NOT be nested in " + name + "; saw:\n" + src,
                src.contains("public enum Priority"));
        }
    }

    @Test
    public void providedEnumIsNotMaterializedAndReferencedAtTheConfiguredNamespace() throws Exception {
        Path out = run(loadShared(), "com.acme.ext");

        // Nothing emitted for the provided Currency type.
        assertFalse("provided Currency must NOT be materialized",
            Files.exists(out.resolve("acme/shop/Currency.java")));

        String ticket = Files.readString(out.resolve("acme/shop/TicketDto.java"));
        assertTrue("currency component must reference the configured external namespace; saw:\n" + ticket,
            ticket.contains("com.acme.ext.Currency currency"));
        assertFalse("provided Currency must NOT be nested in the DTO; saw:\n" + ticket,
            ticket.contains("public enum Currency"));
    }

    @Test
    public void providedEnumWithNoNamespaceConfigIsACodegenErrorNamingTheEnum() throws Exception {
        MetaDataLoader loader = loadShared();
        try {
            run(loader, null); // no providedEnumNamespace
            fail("expected a codegen-time error for the @provided enum with no namespace config");
        } catch (RuntimeException e) {
            assertTrue("error must name the provided enum Currency; was: " + e.getMessage(),
                e.getMessage() != null && e.getMessage().contains("Currency"));
        }
    }
}
